Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,95 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and optional type names that contain
+`public void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
+```
+
+The list is a semicolon-delimited list of assembly paths (absolute, or
+relative to the working directory), each with an optional type name
+(following an exclamation mark). If no type name is specified, the
+default is `StartupHook.Initializer`.
```

Comment:
I am not a fan of magic type names. This is such a low level construct the engineering cost to design a hook grossly outweighs the cost of providing a type name. All it does is create more magic and implicitly creates yet another test case.

Reply:
That's reasonable. I've removed the default type name in the latest iteration.